pkg/reconcilermanager/controllers/reposync_controller.go

```diff
@@ -158,29 +173,104 @@ func (r *RepoSyncReconciler) Reconcile(ctx context.Context, req controllerruntim
 		return controllerruntime.Result{}, updateErr
 	}
 
+	// Create a tombstone, so we know to clean up durring termination.
+	r.addCachedSync(req.Namespace, req.Name)
+
+	finalizerName := v1beta1.SyncFinalizer
+
+	// Cross-namespace owner references are disallowed by Kubernetes.
+	// So we can't rely on Kubernetes' built-in garbage collection to delete
+	// managed objects in the central "config-management-system" namespace, when
+	// the RepoSync is in another namespace.
+	// To avoid leaking managed resources, we use a custom finalizer to force
+	// foreground deletion. This way, the RepoSync will not be fully deleted
+	// until all its managed resources are also deleted.
+
+	if rs.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is NOT being deleted.
+		if !controllerutil.ContainsFinalizer(rs, finalizerName) {
+			// The object is new and doesn't have our finalizer yet.
+			// Add our finalizer and update the object.
+			controllerutil.AddFinalizer(rs, finalizerName)
+			err := r.client.Update(ctx, rs)
+			if err != nil {
+				metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(err), start)
+				return controllerruntime.Result{}, status.APIServerError(err, "failed to patch RepoSync to add finalizer")
+			}
+		}
+	} else {
+		// The object is being deleted.
+		if controllerutil.ContainsFinalizer(rs, finalizerName) {
+			// Our finalizer is present, so delete managed resource objects.
+			if err := r.deleteManagedObjects(ctx, log, reconcilerKey, currentRS, rs, true); err != nil {
+				// Failed to delete a managed object.
+				// Return an error to trigger retry.
+				metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(err), start)
+				return controllerruntime.Result{}, errors.Wrap(err, "failed to delete managed objects")
+			}
+
+			// Remove our finalizer and update the object.
+			controllerutil.RemoveFinalizer(rs, finalizerName)
```

Comment:
Might be good to log that the finalizer successfully completed and is being removed

Reply:
added a few more logs around finalizing, since it shouldn't happen very often.